### ISSUE TITLE
Equality comparison in join should work between comparable types

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#2779]
Repro is with Zed commit [ff3d2e2](https://github.com/brimdata/zed/commit/ff3d2e27bb029836d3fe963a317a19b45f28d14b).

The tl;dr is that at the moment values of different types (specifically `string` and `bstring` in this case) are being treated as non-equal even when their underlying values appear to be the same.

To start with a base case, with a regular test for equality between `string` and `bstring` values, it does what I expect:

```
$ zq -Z stringy.zson 
{
    foo: "hello",
    bar: "hello" (bstring)
} (=0)

$ zq -z 'foo==bar' stringy.zson 
{foo:"hello",bar:"hello" (bstring)} (=0)
```

Now consider this `join`. This next one works because the field called `bar` is a `bstring` in both files.

```
$ zq -Z right.zson 
{
    bar: "hello" (bstring),
    decoration: "LookAtMe!"
} (=0)

$ cat simple.zed 
from (
  file stringy.zson => sort bar;
  file right.zson => sort bar;
) | left join on bar=bar decoration

$ zq -Z -I simple.zed 
{
    foo: "hello",
    bar: "hello" (bstring),
    decoration: "LookAtMe!"
} (=0)
```

However, if I change the type of `bar` in `right.zson` to a regular `string`, now the `join` fails.

```
$ zq -Z right.zson 
{
    bar: "hello",
    decoration: "LookAtMe!"
}

$ zq -Z -I simple.zed 
{
    foo: "hello",
    bar: "hello" (bstring)
} (=0)
```

As a user, the only way I can see to work around this currently is to cast the `bar` value in one of the inputs so the types line up. Once I got hip to the reason the clause in the `join` was failing, this is not a super difficult workaround to put in place, and in a pinch we could surely document it and hope users lock it into their workflow. However, my concern is that users might spend a lot of time frustrated thinking it was some other mistake on their part (e.g. values not quite lining up in some other way, some other problem in the multi-line Zed, etc.)

When I described this to @mccanne he acknowledged that this is something that could be made to work.

**Comment by philrz** [comment]
The Zed docs published in #2836 cited this issue as a current limitation. When this issue is addressed, the docs should be updated.

**Comment by philrz** [comment]
Verified in Zed commit 2ca122e.

Now because the `string` and `bstring` types are comparable, the `join` that was failing above now succeeds (i.e., the `decoration` field is present in the result).

```
$ zq -version
Version: v0.29.0-429-g2ca122e1

$ cat stringy.zson 
{
    foo: "hello",
    bar: "hello" (bstring)
} (=0)

$ cat right.zson 
{
    bar: "hello",
    decoration: "LookAtMe!"
}

$ cat simple.zed 
from (
  file stringy.zson => sort bar;
  file right.zson => sort bar;
) | left join on bar=bar decoration

$ zq -Z -I simple.zed
{
    foo: "hello",
    bar: "hello" (bstring),
    decoration: "LookAtMe!"
} (=0)
```

Thanks @mattnibs!
